# 3.17 - 3.18
- title: Homework 3.17 - 3.18
- permalink: /homework1718

# 3.17 Algorithmic Efficiency

## Vocabulary
- Problem: A general descriptionof a task that can't be solved algorithmically
    - Decision Problem: A problem with a yes or no answer
    - Organization Problem: A problem about finding the best anser
- Instance: A problem with a specific input
- Efficiency: Amount of computing needed for a input
    - Polynomial Efficiency (Good): More work takes a proportional amount of time
    - Exponential Efficiency (Bad): More work takes an exponential amount of time
- Heuristic Approach: When the optimal is inefficient, look for a possibly optimal solution with better efficiency
- Decidable Problem: A problem with a clear answer
- Undecidable Problem: A problem with no solution not been guaranteed to produce the correct the correct output


## Notes
- Less function lead to less efficiency
    - More functions = more time = less efficiency
- Heuristical approach is used to sacrafice the optimal solution for efficiency

## Challenge
Try and fix this ineficcient code!
Only change the code between the two commented lines.
Fully programmed solution will improve your grade, at a minimum show that you tried.

In [3]:
import time
numlist = [1,3,5,7,9,11,13,15,17,19]
valuelist = [0,3,6,9,12,15,18,21]
def isvalue(value,array):
    #--------------------
    return value in array
    """
    Binary Solution (less efficient):
    copy=array.copy()
    copy.sort()
    low=0
    high=len(array)-1
    while low<=high:
        mid = (low+high)//2
        if copy[mid]<value:
            low=mid+1
        else:
            high=mid
    return copy[mid]==value"""
    #--------------------
starttime = time.time()
for i in range(100000):
    for i in range(len(valuelist)):
        x = isvalue(valuelist[i],numlist)
endtime = time.time()
print(endtime-starttime,'seconds') 

0.20411324501037598 seconds


# 3.18 Undecidable Problems

## Notes
- Undecidable problems are results of paradoxes
    - Undecidable problems halts progress.

# Homework!

Make an algorithm that finds the fastest route that hits every location once starting and ending at Del Norte. Make sure to show your thinking. If you are strugling, try using a huristic approach. Remember, what matters more than having perfectly functioning code is that you tried your hardest.

In [54]:
# RUN THIS CODE --- DO NOT EDIT
dataset = {
    'DelNorte':{
        'Westview':15,
        'MtCarmel':20,
        'Poway':35,
        'RanchoBernrdo':50
    },
    'Westview':{
        'Del Norte':15,
        'MtCarmel':35,
        'Poway':25,
        'RanchoBernrdo': 45
    },
    'MtCarmel':{
        'Westview':35,
        'Del Norte':20,
        'Poway':40,
        'RanchoBernrdo':30
    },
    'Poway':{
        'Westview':25,
        'MtCarmel':40,
        'Del Norte':35,
        'RanchoBernrdo':15
    },
    'RanchoBernardo':{
        'Westview':45,
        'MtCarmel':30,
        'Poway':15,
        'Del Norte':50
    }
}

In [ ]:
def fastestroute(start,data):
    drivetime = 0
    order = []
    #CODE,CODE,CODE
    return(drivetime,order)

start = 'DelNorte'
# 'dataset' is the name of the nested key value pair

### Grading:
|Challenge|Homework|
|---|---|
|.15 pts for attempt|.65 for attempt|
|.20 pts for complete|.70 for complete|
|.25 pts for above and beyond|.75 pts for above and beyond|
